test/core/bad_connection/close_fd_test.cc

```diff
@@ -0,0 +1,731 @@
+  GPR_ASSERT(ev.tag == tag(1));
+  GPR_ASSERT(status == GRPC_STATUS_UNAVAILABLE);
+
+  /* No event is received on the server */
+  ev = grpc_completion_queue_next(
+      ctx.cq, grpc_timeout_milliseconds_to_deadline(100), nullptr);
+  GPR_ASSERT(ev.success == 0);
+  GPR_ASSERT(ev.type == GRPC_QUEUE_TIMEOUT);
+  GPR_ASSERT(ev.tag == nullptr);
+
+  grpc_slice_unref(details);
+  grpc_metadata_array_destroy(&initial_metadata_recv);
+  grpc_metadata_array_destroy(&trailing_metadata_recv);
+  grpc_metadata_array_destroy(&request_metadata_recv);
+  grpc_call_details_destroy(&call_details);
+
+  grpc_call_unref(c);
+
+  grpc_byte_buffer_destroy(request_payload);
+  grpc_byte_buffer_destroy(response_payload);
+  grpc_byte_buffer_destroy(request_payload_recv);
+  grpc_byte_buffer_destroy(response_payload_recv);
+
+  end_test(&ctx);
+}
+static void test_close_before_client_send() {
+  /* Close client side of the connection before client sends message to server
+   * */
+  _test_close_before_client_send(CLIENT_FD);
+  /* Close server side of the connection before client sends message to server
+   * */
+  _test_close_before_client_send(SERVER_FD);
+}
+
+static void _test_close_before_call_create(fd_type f) {
+  grpc_core::ExecCtx exec_ctx;
+  grpc_call* c;
+  grpc_event ev;
+  test_ctx ctx;
+  test_init(&ctx);
+
+  gpr_timespec deadline = grpc_timeout_milliseconds_to_deadline(100);
+  grpc_endpoint_add_to_pollset(ctx.ep->client, grpc_cq_pollset(ctx.client_cq));
+
+  grpc_endpoint_pair* sfd = ctx.ep;
+  int fd;
+  if (f == SERVER_FD) {
+    fd = sfd->server->vtable->get_fd(sfd->server);
+  } else {
+    GPR_ASSERT(f == CLIENT_FD);
+    fd = sfd->client->vtable->get_fd(sfd->client);
+  }
+  /* Connection is closed before the client creates a call */
+  close(fd);
+
+  c = grpc_channel_create_call(
+      ctx.client, nullptr, GRPC_PROPAGATE_DEFAULTS, ctx.client_cq,
+      grpc_slice_from_static_string("/foo"), nullptr, deadline, nullptr);
+  GPR_ASSERT(c);
+
+  /* Client and server time out waiting on their completion queues and nothing
+   * is sent or received */
+  ev = grpc_completion_queue_next(
+      ctx.client_cq, grpc_timeout_milliseconds_to_deadline(100), nullptr);
+  GPR_ASSERT(ev.type == GRPC_QUEUE_TIMEOUT);
+  GPR_ASSERT(ev.success == 0);
+  GPR_ASSERT(ev.tag == nullptr);
+
+  ev = grpc_completion_queue_next(
+      ctx.cq, grpc_timeout_milliseconds_to_deadline(100), nullptr);
+  GPR_ASSERT(ev.type == GRPC_QUEUE_TIMEOUT);
+  GPR_ASSERT(ev.success == 0);
+  GPR_ASSERT(ev.tag == nullptr);
+
+  grpc_call_unref(c);
+  end_test(&ctx);
+}
+
+static void test_close_before_call_create() {
+  /* Close client side of the connection before client creates a call */
+  _test_close_before_call_create(CLIENT_FD);
+  /* Close server side of the connection before client creates a call */
+  _test_close_before_call_create(SERVER_FD);
+}
+
+int main(int argc, char** argv) {
+  grpc::testing::TestEnvironment env(argc, argv);
+  /* Init grpc */
+  grpc_init();
+
```

Comment:
How about a loop that tests each case for multiple times? I feel that under common case it should be fine but sometimes failures/crashes happen depending on threading. (This maybe better to be done in a separate task like the performance tests for C++; I am not very sure right now).